jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/WebSocketChannel.java

```diff
@@ -389,43 +399,44 @@ public void onOpen()
         if (LOG.isDebugEnabled())
             LOG.debug("onOpen() {}", this);
 
-        try
-        {
-            // Upgrade success
-            channelState.onConnected();
+        // Upgrade success
+        channelState.onConnected();
+        if (LOG.isDebugEnabled())
+            LOG.debug("ConnectionState: Transition to CONNECTED");
 
-            if (LOG.isDebugEnabled())
-                LOG.debug("ConnectionState: Transition to CONNECTED");
+        Callback openCallback =  Callback.from(()->
+                {
+                    channelState.onOpen();
+                    if (!demanding)
+                        connection.demand(1);
```

Comment:
I am dubious this is correct. 
Field `demanding` is only ever assigned in the constructor, copying the value of the handler which at this point has just been created by the application - so it's a fixed value at the moment of creation.
IMHO it is better for the application to signal if it wants demand by calling `coreSession.demand(1)` from `onOpen(session, callback)`. If the application does not demand, it will have arranged to do so later (e.g. in a WebSocket proxy it will do it after having opened a connection to the upstream server).
I would remove the `WebSocketChannel.demanding` field.

Reply:
I don't think you have understood the `demanding`  field.  Perhaps it should be called `notAutoDemand`.
If `demanding` is false, then the application is saying that it does not want to manage demand, so core handles demand on it's behalf - with an implicit demand(1) done on every successful callback completion. This is the 9.4 behaviour.
If `demanding` is true, then core will never call demand and the application must handle demand itself.  This is the new behaviour that allows back pressure to be handled independently of callbacks.

Reply:
I don't think core should do this. We will always have an upper layer implementation, being JSR, Jetty or Reactive that knows exactly how to do it based on the API it exposes.
So IMHO core should always _not_ automatically demand, and the upper layer implementations should do the right thing. 
JSR has no API for async demand, so it will always demand once callbacks are succeeded.
Jetty and Reactive will expose API to demand.
Core is low level so whoever programs against the core API should know what to do - it goes towards a unique mechanism instead of 2 and overall simplification.

Reply:
I think requiring users of core to manage demand when they do not wish to do so is error prone, as frameworks are likely to not cover some exceptional cases.   It can be done simply and trivially in core with good testing.  We currently have two uses of core, neither of which manage demand, so I think that adding the new feature of demand should be optional.